### PR TITLE
Ci enable distconv

### PR DESCRIPTION
**Pull request by bvanessen**
Added new CI tests for building and checking execution with both DistConv and NVSHMEM enabled as appropriate.

### CODE REVIEW

File: ci_test/common_python/tools.py [hunk 1]
```diff
@@ -1022,4 +1022,5 @@ def gpus_per_node(lbann):
 def get_distconv_environment():
     # TODO: Use the default halo exchange and shuffle method. See https://github.com/LLNL/lbann/issues/1659
     return {"LBANN_DISTCONV_HALO_EXCHANGE": "AL",
-            "LBANN_DISTCONV_TENSOR_SHUFFLER": "AL"}
+            "LBANN_DISTCONV_TENSOR_SHUFFLER": "AL",
+            "LBANN_KEEP_ERROR_SIGNALS" : "1"}
```
Comment (tbennun) [inline ci_test/common_python/tools.py:1026]: Why is this necessary?

**Review by tbennun — COMMENTED** [review]
Looks good, one question about the code.

File: ci_test/common_python/tools.py (hunk 1 above)
Comment (bvanessen) [reply ci_test/common_python/tools.py:1026]: @tbennun Because most of the DistConv layers will fail when those signals are not kept.  I believe that this is an issue with respect to how data is moved between the DistConv and normal tensors. @benson31 Can you comment more.

Comment (benson31) [reply ci_test/common_python/tools.py:1026]: The short version is that distconv adapters don't maintain various layer invariants (yeah, it's a problem) and this is a workaround to make sure necessary data is kept around.

Comment (tbennun) [reply ci_test/common_python/tools.py:1026]: Thanks

**Review by tbennun — APPROVED** [review]

File: .gitlab-ci.yml [hunk 2]
```diff
@@ -49,6 +62,19 @@ lassen testing:
...
+#   variables:
+#     JOB_NAME_SUFFIX: _distconv
+#     SPACK_ENV_BASE_NAME_MODIFIER: "-multi-stage-distconv"
+#     SPACK_SPECS: "+cuda +distconv +nvshmem +fft"
+#     WITH_WEEKLY: "${LBANN_CI_RUN_WEEKLY}"
+#     WITH_CLEAN_BUILD: "${LBANN_CI_CLEAN_BUILD}"
+#     TEST_FLAG: "test_*_distconv.py"
+#   trigger:
+#     strategy: depend
+#     include: .gitlab/lassen/multi_stage_pipeline.yml
```
Comment (benson31) [inline .gitlab-ci.yml:76]: What's the issue here? Arguably this is the most important of the test jobs for this PR...

File: python/lbann/contrib/lc/launcher.py [hunk 3]
```diff
@@ -91,8 +94,10 @@ def prepend_environment_path(key, prefix):
     # Optimizations for Tioga
     if system in ('tioga', 'rzvernal'):
         #set_environment('NCCL_SOCKET_IFNAME', 'hsi')
-        set_environment('MIOPEN_DEBUG_DISABLE_FIND_DB', '1')
-        set_environment('MIOPEN_DISABLE_CACHE', '1')
+        set_environment('MIOPEN_DEBUG_DISABLE_FIND_DB', '0')
+        set_environment('MIOPEN_DISABLE_CACHE', '0')
+        set_environment('MIOPEN_USER_DB_PATH', '/tmp/MIOpen_user_db')
+        set_environment('MIOPEN_CUSTOM_CACHE_DIR', '/tmp/MIOpen_custom_cache')
```
Comment (benson31) [inline python/lbann/contrib/lc/launcher.py:100]:
```suggestion
        set_environment('MIOPEN_USER_DB_PATH', '${TMPDIR}/MIOpen_user_db')
        set_environment('MIOPEN_CUSTOM_CACHE_DIR', '${TMPDIR}/MIOpen_custom_cache')
```

File: python/lbann/contrib/lc/launcher.py [hunk 4]
```diff
@@ -117,8 +122,10 @@ def prepend_environment_path(key, prefix):
         set_environment('OMPI_MCA_mpi_warn_on_fork', 0)
 
         #set_environment('NCCL_SOCKET_IFNAME', 'hsi')
-        set_environment('MIOPEN_DEBUG_DISABLE_FIND_DB', '1')
-        set_environment('MIOPEN_DISABLE_CACHE', '1')
+        set_environment('MIOPEN_DEBUG_DISABLE_FIND_DB', '0')
+        set_environment('MIOPEN_DISABLE_CACHE', '0')
+        set_environment('MIOPEN_USER_DB_PATH', '/tmp')
+        set_environment('MIOPEN_CUSTOM_CACHE_DIR', '/tmp')
```
Comment (benson31) [inline python/lbann/contrib/lc/launcher.py:128]:
```suggestion
        set_environment('MIOPEN_USER_DB_PATH', '${TMPDIR}/MIOpen_user_db')
        set_environment('MIOPEN_CUSTOM_CACHE_DIR', '${TMPDIR}/MIOpen_custom_cache')
```
For consistency.

File: python/lbann/contrib/olcf/launcher.py [hunk 5]
```diff
@@ -103,8 +103,8 @@ def prepend_environment_path(key, prefix):
         #set_environment('NCCL_SOCKET_IFNAME', 'hsi')
         set_environment('MIOPEN_DEBUG_DISABLE_FIND_DB', '1')
         set_environment('MIOPEN_DISABLE_CACHE', '1')
-        set_environment('MIOPEN_USER_DB_PATH', '/tmp')
-        set_environment('MIOPEN_CUSTOM_CACHE_DIR', '/tmp')
+        set_environment('MIOPEN_USER_DB_PATH', '/tmp/MIOpen_user_db')
+        set_environment('MIOPEN_CUSTOM_CACHE_DIR', '/tmp/MIOpen_custom_cache')
```
Comment (benson31) [inline python/lbann/contrib/olcf/launcher.py:107]:
```suggestion
        set_environment('MIOPEN_USER_DB_PATH', '${TMPDIR}/MIOpen_user_db')
        set_environment('MIOPEN_CUSTOM_CACHE_DIR', '${TMPDIR}/MIOpen_custom_cache')
```

File: python/lbann/launcher/flux.py [hunk 6]
```diff
@@ -121,15 +121,16 @@ def add_parallel_command(self,
...
-        args.append(f'-g 1') # --gpus-per-task
+#        args.append(f'-g 1') # --gpus-per-task
         # Ramesh had used a -c flag but doesn't  seem to use it right now
         # args.append(f'-c {int(self.cores_per_node / procs_per_node)}') #--cores-per-task
-        args.append(f'-o gpu-affinity=per-task')
-        args.append(f'-o cpu-affinity=per-task')
+#        args.append(f'-o gpu-affinity=per-task')
+#        args.append(f'-o cpu-affinity=per-task')
         args.append(f'-o nosetpgrp')
+#        args.append(f'-o pmi=pmix')
```
Comment (benson31) [inline python/lbann/launcher/flux.py:133]:
```suggestion
```
This has moved into LC-specific stuff.

File: python/lbann/launcher/flux.py [hunk 7]
```diff
@@ -121,15 +121,16 @@ def add_parallel_command(self,
         args.append(f'--setattr=system.cwd={work_dir}')
         args.append(f'--nodes={nodes}')
         args.append(f'--ntasks={nodes * procs_per_node}')
-        args.append(f'-o per-resource.type=node')
-        args.append(f'-o per-resource.count={procs_per_node}')
+#        args.append(f'-o per-resource.type=node')
+#        args.append(f'-o per-resource.count={procs_per_node}')
```
Comment (benson31) [inline python/lbann/launcher/flux.py:125]:
```suggestion
```

File: python/lbann/launcher/flux.py [hunk 8]
```diff
@@ -121,15 +121,16 @@ def add_parallel_command(self,
...
         args.append(f'--exclusive')
-        args.append(f'-g 1') # --gpus-per-task
+#        args.append(f'-g 1') # --gpus-per-task
```
Comment (benson31) [inline python/lbann/launcher/flux.py:127]:
```suggestion
```

File: python/lbann/launcher/flux.py [hunk 9]
```diff
@@ -121,15 +121,16 @@ def add_parallel_command(self,
...
         # Ramesh had used a -c flag but doesn't  seem to use it right now
         # args.append(f'-c {int(self.cores_per_node / procs_per_node)}') #--cores-per-task
-        args.append(f'-o gpu-affinity=per-task')
-        args.append(f'-o cpu-affinity=per-task')
+#        args.append(f'-o gpu-affinity=per-task')
+#        args.append(f'-o cpu-affinity=per-task')
```
Comment (benson31) [inline python/lbann/launcher/flux.py:131]:
```suggestion
```

File: include/lbann/data_coordinator/data_coordinator.hpp [hunk 10]
```diff
@@ -281,7 +281,8 @@ class data_coordinator
 
   bool at_new_epoch() const;
 
-  virtual void register_active_data_field(data_field_type const data_field);
+  virtual void register_active_data_field(data_field_type const data_field,
+                                          std::vector<El::Int> const data_field_dim_map);
```
Comment (benson31) [inline include/lbann/data_coordinator/data_coordinator.hpp:285]: Do you mean
```
  virtual void register_active_data_field(data_field_type const& data_field,
                                          std::vector<El::Int> const& data_field_dim_map);
```
? Generally, C++ does not declare "pass-by-value" parameters to be `const`. I would suggest either removing the `const`s or making these "pass-by-const-reference" parameters.

File: src/data_coordinator/data_coordinator.cpp [hunk 11]
```diff
@@ -349,6 +350,15 @@ long data_coordinator::get_linearized_size(
       linearized_size = tmp_size;
     }
   }
+  auto& dim_map = m_active_data_fields_dim_map.at(data_field);
+  if (linearized_size != get_linear_size(dim_map)) {
+    if (linearized_size == -1) {
+      LBANN_WARNING("Unable to find data readers; using data field map for linearized size for data field: ", data_field, " = ", get_linear_size(dim_map));
+      linearized_size = get_linear_size(dim_map);
+    }else {
```
Comment (benson31) [inline src/data_coordinator/data_coordinator.cpp:358]:
```suggestion
    }
    else {
```
clang-format?

File: src/utils/lbann_library.cpp [hunk 12]
```diff
@@ -53,7 +53,7 @@ namespace lbann {
 
 // Creates a datareader metadata to get around the need for an actual
 // datareader in inference only mode
-auto mock_dr_metadata(std::vector<int> input_dims, std::vector<int> output_dims)
+  auto mock_dr_metadata(std::vector<El::Int> input_dims, std::vector<El::Int> output_dims)
```
Comment (benson31) [inline src/utils/lbann_library.cpp:56]:
```suggestion
auto mock_dr_metadata(std::vector<El::Int> input_dims, std::vector<El::Int> output_dims)
```

File: ci_test/unit_tests/test_unit_layer_pooling_distconv.py [hunk 13]
```diff
@@ -310,10 +312,10 @@ def construct_data_reader(lbann):
...
 # Setup PyTest
 # ==============================================
-
 # Create test functions that can interact with PyTest
 # Note: Create test name by removing ".py" from file name
 _test_name = os.path.splitext(os.path.basename(current_file))[0]
 for _test_func in tools.create_tests(setup_experiment, _test_name,
-                               environment=tools.get_distconv_environment()):
+                                     environment=lbann.contrib.args.get_distconv_environment(),
+                                     time_limit=15):
```
Comment (benson31) [inline ci_test/unit_tests/test_unit_layer_pooling_distconv.py:320]: Oof. If that's what it has to be, that's what it has to be. But ouch.

File: ci_test/unit_tests/test_unit_layer_convolution_distconv.py [hunk 14]
```diff
@@ -292,5 +293,7 @@ def construct_data_reader(lbann):
 # Note: Create test name by removing ".py" from file name
 _test_name = os.path.splitext(os.path.basename(current_file))[0]
 for _test_func in tools.create_tests(setup_experiment, _test_name,
-                               environment=tools.get_distconv_environment()):
+                                     environment=lbann.contrib.args.get_distconv_environment(),
+                                     skip_clusters=["tioga"],
```
Comment (benson31) [inline ci_test/unit_tests/test_unit_layer_convolution_distconv.py:297]: This is already tracked in #2326. It might be good to add more detail to that ticket regarding the manner of failure.

File: ci_test/integration_tests/test_integration_resnet50.py [hunk 15]
```diff
@@ -166,7 +166,7 @@ def augment_test_func(test_func):
     def func(cluster, dirname, weekly):
 
         # Skip test on CPU systems
-        if cluster in ('catalyst', 'corona'):
+        if cluster in ('catalyst'):
```
Comment (benson31) [inline ci_test/integration_tests/test_integration_resnet50.py:169]: We can test `lbann.has_feature('GPU')`. RIP Catalyst.

File: ci_test/common_python/tools.py [hunk 16]
```diff
@@ -744,6 +747,30 @@ def create_tests(setup_func,
...
+
+        if os.environ.get('USER') == 'lbannusr':
+            basepath = '/p/vast1/lbannusr'
+        else:
+            basepath = '/p/vast1/lbann'
+
+        if os.path.isdir(basepath) and os.access(basepath, os.R_OK | os.W_OK):
+            db_path = basepath
+        else:
+            db_path = '/tmp'
```
Comment (benson31) [inline ci_test/common_python/tools.py:766]:
```suggestion
            db_path = '${TMPDIR}'
```

File: ci_test/common_python/tools.py [hunk 17]
```diff
@@ -744,6 +747,30 @@ def create_tests(setup_func,
...
+
+        environment['MIOPEN_USER_DB_PATH'] = f'{db_path}/MIOpen_user_db'
+        # Empirically the cache dir cannot be on a parallel file system
+        environment['MIOPEN_CUSTOM_CACHE_DIR'] ='/tmp/MIOpen_custom_cache'
```
Comment (benson31) [inline ci_test/common_python/tools.py:770]:
```suggestion
        environment['MIOPEN_CUSTOM_CACHE_DIR'] ='${TMPDIR}/MIOpen_custom_cache'
```

**Review by benson31 — DISMISSED** [review]
Should be quick comments.

File: include/lbann/utils/dim_helpers.hpp [hunk 18]
```diff
@@ -99,6 +99,12 @@ auto get_packed_strides(std::vector<T> const& dims)
   return get_strides(dims.size(), dims.data(), T(1));
 }
 
+template <typename To, typename From>
+auto vector_cast(std::vector<From> const& from)
+{
+  return std::vector<To>{from.cbegin(), from.cend()};
+}
+
```
Comment (benson31) [inline include/lbann/utils/dim_helpers.hpp:107]: Flagging this for future work: It would be useful if we had a narrow-safe version of this to use in debug builds.
